File: src/tree/c/bsTree.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "binaryTree.h"

struct BinaryTree* createBinaryTree(){
    struct BinaryTree *tree = malloc(sizeof(struct BinaryTree));
    tree->root = NULL;
    tree->nodeCount = 0;

    return tree;
}

int insert(struct BinaryTree *tree, int id, char *data){
    struct Node *leaf = search(id, tree->root);

    if (leaf && id == leaf->id){
        printf("Node with id '%d' already exists\n", id);
        return -1;
    }

    struct Node *node = malloc(sizeof(struct Node));
    node->left = NULL;
    node->right = NULL;
    node->id = id;
    node->data = malloc(sizeof(char) * (strlen(data) + 1));
    strncpy(node->data, data, strlen(data) + 1);

    tree->nodeCount += 1;

    if (leaf == NULL){
        tree->root = node;
        return 0;
    }

    if (id > leaf->id){
        leaf->right = node;
    } else {
        leaf->left = node;
    }

    return 0;
}

int traverse(struct BinaryTree *tree, enum TraverseStrategy strategy){
    if (isEmpty(tree)){
        printf("Tree is empty\n");
        return -1;
    }

    if (strategy == 0){
        levelTraversal(tree);
        printf("\n");
    } else {
        depthTraversal(tree->root, strategy);
        printf("\n");
    }

    return 0;
}

int delete(struct BinaryTree *tree, int id){
    if (isEmpty(tree)){
        printf("Tree is empty\n");
        return -1;
    }

    tree->root = deleteNode(tree, tree->root, id);
    return 0;

}

int update(struct BinaryTree *tree, int id, char* data) {
    if (isEmpty(tree)){
        printf("Tree is empty\n");
        return -1;
    }

    struct Node *foundNode = search(id, tree->root);

    if (id != foundNode->id){
        printf("Node with id '%d' not found\n", id);
        return -1;
    }

    free(foundNode->data);
    foundNode->data = malloc(sizeof(char) * strlen(data) + 1);
    strncpy(foundNode->data, data, strlen(data) + 1);

    return 0;
}

char* find(int id, struct BinaryTree *tree){
    if (isEmpty(tree)){
        printf("Tree is empty\n");
        return NULL;
    }

    struct Node *foundNode = search(id, tree->root);

    if (id != foundNode->id){
        printf("Node with id '%d' not found\n", id);
        return NULL;
    }

    return foundNode->data;
}

char* min(struct BinaryTree *tree){
    if (isEmpty(tree)){
        printf("Tree is empty\n");
        return NULL;
    }

    struct Node *current = tree->root;

    while (current->left != NULL){
        current = current->left;
    }

    return current->data;

}

char* max(struct BinaryTree *tree){
    if (isEmpty(tree)){
        printf("Tree is empty\n");
        return NULL;
    }

    struct Node *current = tree->root;

    while (current->right != NULL){
        current = current->right;
    }

    return current->data;

}

int isEmpty(struct BinaryTree *tree){
    return tree->root == NULL;
}

static struct Node* search(int id, struct Node *root){
    if (root && id > root->id){
        return root->right == NULL ? root: search(id, root->right);
    } else if (root && id < root->id){
        return root->left == NULL ? root: search(id, root->left);
    } else {
        return root;
    }
}

static struct Node* deleteNode(struct BinaryTree *tree, struct Node *node, int id){
    if (node == NULL){
        printf("Node with id '%d' not found\n", id);
        return NULL;
    }

    if (id < node->id){
        node->left = deleteNode(tree, node->left, id);
    } else if (id > node->id){
        node->right = deleteNode(tree, node->right, id);
    } else{
        tree->nodeCount -= 1;

        if (node->left == NULL && node->right == NULL){
            free(node->data);
            free(node);
            return NULL;

        } else if (!(node->left && node->right)){
            struct Node *temp = node->left ? node->left : node->right;
            free(node->data);
            free(node);
            return temp;

        } else {
            struct Node *current = node->left;
            struct Node *previous = NULL;

            while (current->right != NULL){
                previous = current;
                current = current->right;
            }

            free(node->data);
            node->id = current->id;
            node->data = malloc(sizeof(char) * strlen(current->data) + 1);
            strncpy(node->data, current->data, strlen(current->data) + 1);

            if (previous){
                previous->right = current->left ? current->left : current->right;
            } else {
                node->left = current->left ? current->left : current->right;
            }

            free(current->data);
            free(current);


            return node;
        }
    }

    return node;
}

static int depthTraversal(struct Node *root, enum TraverseStrategy strategy){
    switch (strategy) {
        case 1:
            printf("[id: %d, data: %s]\n", root->id, root->data);
            root->left != 0? depthTraversal(root->left, strategy): 0;
            root->right != 0? depthTraversal(root->right, strategy): 0;
            break;
        case 2:
            root->left != 0? depthTraversal(root->left, strategy): 0;
            printf("[id: %d, data: %s]\n", root->id, root->data);
            root->right != 0? depthTraversal(root->right, strategy): 0;
            break;
        case 3:
            root->left != 0? depthTraversal(root->left, strategy): 0;
            root->right != 0? depthTraversal(root->right, strategy): 0;
            printf("[id: %d, data: %s]\n", root->id, root->data);
            break;
        default:
            printf("Traverse strategy not supported\n");
            return -1;
    }

    return 0;
}

static int levelTraversal(struct BinaryTree *tree){
    struct Node *current;
    struct Node **queue = malloc(sizeof(struct Node*) * tree->nodeCount);
    int begin = 0;
    int end = 0;

    queue[end] = tree->root;

    while (begin < tree->nodeCount){
        if (queue[begin] != NULL){
            printf("[id: %d, data: %s]\n", queue[begin]->id, queue[begin]->data);
        }

        current = queue[begin];
        if (current->left != NULL){
            end += 1;
            queue[end] = current->left;
        }
        if (current->right != NULL){
            end += 1;
            queue[end] = current->right;
        }
        begin += 1;
    }

    free(queue);
    return 0;
}

File: src/dataStructures/tree/ts/bsTree.ts
import { BinaryTree, INode, TraverseStrategy } from "./types"
import { EmptyTreeError, NotFoundError, RepeatedNodeError } from "./utils"

export class Node<Type> implements INode<Type> {
    public id: number
    public data: Type
    public left: INode<Type> | null
    public right: INode<Type> | null

    constructor(id: number, data: Type) {
        this.id = id
        this.data = data
        this.left = null
        this.right = null
    }
}

export class BinarySearchTree<Type> implements BinaryTree<Type> {
    protected root: INode<Type> | null

    constructor() {
        this.root = null
    }

    public insert(id: number, data: Type): void {
        const newNode = new Node(id, data)

        if (this.root === null) {
            this.root = newNode
            return
        }

        const leaf = this.search(id, this.root)

        if (id > leaf.id) {
            leaf.right = newNode
        } else if (id < leaf.id) {
            leaf.left = newNode
        } else {
            throw new RepeatedNodeError(id)
        }
    }

    public find(id: number): Pick<INode<Type>, "id" | "data"> {
        if (this.root === null) {
            throw new EmptyTreeError()
        }

        const foundNode = this.search(id, this.root)

        if (foundNode.id !== id) {
            throw new NotFoundError(id)
        }

        return { id: foundNode.id, data: foundNode.data }
    }

    public delete(id: number): void {
        if (this.root === null) {
            throw new EmptyTreeError()
        }

        this.root = this.deleteNode(this.root, id)
    }

    public update(id: number, data: Type): Pick<INode<Type>, "id" | "data"> {
        if (this.root === null) {
            throw new EmptyTreeError()
        }

        const foundNode = this.search(id, this.root)

        if (id !== foundNode.id) {
            throw new NotFoundError(id)
        }

        foundNode.data = data
        return { id, data }
    }

    public traverse(strategy: TraverseStrategy): void {
        if (this.root === null) {
            throw new EmptyTreeError()
        }

        if (strategy === "level") {
            this.levelTraversal(this.root)
        } else {
            this.depthTraversal(this.root, strategy)
        }
    }

    public min(): Pick<INode<Type>, "id" | "data"> {
        if (this.root === null) {
            throw new EmptyTreeError()
        }

        let current = this.root

        while (current.left !== null) {
            current = current.left
        }

        return { id: current.id, data: current.data }
    }

    public max(): Pick<INode<Type>, "id" | "data"> {
        if (this.root === null) {
            throw new EmptyTreeError()
        }

        let current = this.root

        while (current.right !== null) {
            current = current.right
        }

        return { id: current.id, data: current.data }
    }

    public print(): void {
        console.log(JSON.stringify(this.root, null, 2))
    }

    public isEmpty(): boolean {
        return this.root === null
    }

    private search(id: number, root: INode<Type>): INode<Type> {
        if (id > root.id) {
            return root.right === null ? root : this.search(id, root.right)
        } else if (id < root.id) {
            return root.left === null ? root : this.search(id, root.left)
        } else {
            return root
        }
    }

    private depthTraversal(root: INode<Type>, strategy: TraverseStrategy): void {
        if (strategy === "preOrder") {
            console.log(JSON.stringify({ id: root.id, data: root.data }))
        }
        root.left && this.depthTraversal(root.left, strategy)
        if (strategy === "inOrder") {
            console.log(JSON.stringify({ id: root.id, data: root.data }))
        }
        root.right && this.depthTraversal(root.right, strategy)
        if (strategy === "postOrder") {
            console.log(JSON.stringify({ id: root.id, data: root.data }))
        }
    }

    private levelTraversal(root: INode<Type>): void {
        const queue: INode<Type>[] = [root]
        let currentNode = root
        let pointer = 0

        while (pointer < queue.length) {
            console.log({ id: queue[pointer].id, data: queue[pointer].data })

            currentNode.left && queue.push(currentNode.left)
            currentNode.right && queue.push(currentNode.right)

            pointer += 1
            currentNode = queue[pointer]
        }
    }

    private deleteNode(root: INode<Type> | null, id: number): INode<Type> | null {
        if (root === null) {
            throw new NotFoundError(id)
        }

        if (id < root.id) {
            root.left = this.deleteNode(root.left, id)
        } else if (id > root.id) {
            root.right = this.deleteNode(root.right, id)
        } else {
            if (!root.left && !root.right) {
                return null
            } else if (!(root.right && root.left)) {
                return root.right || root.left
            } else {
                const { data, id } = this.max.call({ root: root.left })
                root.left = this.deleteNode(root.left, id)
                root.id = id
                root.data = data
            }
        }

        return root
    }
}
